### PR TITLE
[DPE-1720] Add backups to k8s charm and configure PBM

### PR DESCRIPTION
**Pull request by dmitry-ratushnyy**
## About

The goal of this PR is to
-  add backup library to MongoDB K8S charm 
- configure PBM
- Add unit and integration tests for PBM configuration

## Related issues
[DPE-1720], [DPE-1957]

[DPE-1720]: https://warthogs.atlassian.net/browse/DPE-1720?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DPE-1957]: https://warthogs.atlassian.net/browse/DPE-1957?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lib/charms/data_platform_libs/v0/s3.py
Comment (dmitry-ratushnyy) [inline lib/charms/data_platform_libs/v0/s3.py:1]: This file is copied from VM charm

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+    def substrate(self) -> str:
+        """Returns the substrate of the charm."""
+        return self._substrate
+
+    def _on_s3_credential_changed(self, event: CredentialsChangedEvent):
+        """Sets pbm credentials, resyncs if necessary and reports config errors."""
+        # handling PBM configurations requires that MongoDB is running and the pbm snap is
+        # installed.
+        if not self.charm.db_initialised:
+            logger.error("Cannot set PBM configurations, MongoDB has not yet started.")
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:105]: I think it is `info` maximum
```suggestion
            logger.info("Defering: Cannot set PBM configurations, MongoDB has not yet started.")
```

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+        try:
+            # TODO sync with VM lib
+            if self.substrate() == "k8s":
+                self.charm.get_backup_service()
+        except ModelError:
+            logger.error("Cannot set PBM configurations, pbm-agent service not found.")
+            event.defer()
+        except RuntimeError:
+            logger.error("Cannot set PBM configurations. Failed to get pbm serivice.")
+            self.charm.unit.status = BlockedStatus("couldn't configure s3 backup options.")
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:118]:
```suggestion
        try:
            self.charm.get_backup_service()
        except (ModelError, RuntimeError) as e:
            logger.info("Defering: Cannot set PBM configurations, pbm-agent service not found. error: %s.", str(e))
            event.defer()
            return
```

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+    # BEGIN: helper functions
+
+    def _configure_pbm_options(self, event) -> None:
+        try:
+            self._set_config_options(self._get_pbm_configs())
+            self._resync_config_options()
+        except SetPBMConfigError:
+            self.charm.unit.status = BlockedStatus("couldn't configure s3 backup options.")
+            return
+        except RuntimeError as e:  # TODO on VM charm here should be snap.SnapError
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:218]:
```suggestion
        except (snap.SnapError, RuntimeError) as e:
```

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+            logger.error("An exception occurred when starting pbm agent, error: %s.", str(e))
+            self.charm.unit.status = BlockedStatus("couldn't start pbm")
+            return
+        except ResyncError:
+            self.charm.unit.status = WaitingStatus("waiting to sync s3 configurations.")
+            event.defer()
+            logger.error("Sync-ing configurations needs more time.")
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:225]:
```suggestion
            logger.info("Defering: Sync-ing configurations needs more time.")
```

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+            return
+        except PBMBusyError:
+            self.charm.unit.status = WaitingStatus("waiting to sync s3 configurations.")
+            logger.error(
+                "Cannot update configs while PBM is running, must wait for PBM action to finish."
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:230]:
```suggestion
            logger.info(
                    "Defering: Cannot update configs while PBM is running, must wait for PBM action to finish."
```

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+                )
+            except ExecError as e:
+                logger.error(f"Failed to set pbm config file. {e}")
+                self.charm.unit.status = BlockedStatus(self._process_pbm_error(e))
+            return
+        else:
+            subprocess.check_output(
+                f"charmed-mongodb.pbm config --file {MONGODB_SNAP_DATA_DIR}/etc/pbm/pbm_config.yaml",
+                shell=True,
+            )
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:278]:
```suggestion
    def _set_pbm_config_file(self) -> None:
        try:
            self.charm.write_file(
                PBM_CONFIG_FILE_PATH_K8S,
                "# this file is to be left empty. Changes in this file will be ignored.\n",
                make_dirs=True,
                permissions=0o400,
            )
            self.charm.run_pbm_command(
                [PBM_PATH, "config", "--file", PBM_CONFIG_FILE_PATH_K8S]
            )
        except ExecError as e:
            logger.error(f"Failed to set pbm config file. {e}")
            self.charm.unit.status = BlockedStatus(self._process_pbm_error(e))
        return
```

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+            if s3_option not in S3_PBM_OPTION_MAP:
+                continue
+
+            pbm_configs[S3_PBM_OPTION_MAP[s3_option]] = s3_value
+        return pbm_configs
+
+    def _resync_config_options(self):
+        """Attempts to sync pbm config options and sets status in case of failure."""
+        container = self.charm.get_container()
+        container.start("pbm-agent")
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:307]:
```suggestion
        self.charm.service("start", "pbm-agent")
```

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+        ):
+            with attempt:
+                pbm_status = self._get_pbm_status()
+                # wait for backup/restore to finish
+                if isinstance(pbm_status, (MaintenanceStatus)):
+                    raise PBMBusyError
+
+                # if a resync is running restart the service
+                if isinstance(pbm_status, (WaitingStatus)):
+                    container.restart("pbm-agent")
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:324]:
```suggestion
                    self.charm.service("restart", "pbm-agent")
```

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+            with attempt:
+                try:
+                    if self.substrate() == "k8s":
+                        pbm_status = self.charm.run_pbm_command(PBM_STATUS_CMD)
+                    else:
+                        pbm_status = subprocess.check_output(
+                            "charmed-mongodb.pbm status", shell=True
+                        )
+                        pbm_status = pbm_status.decode("utf-8")
+
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:368]:
```suggestion
                        pbm_status = self.charm.run_command(PBM_STATUS_CMD)

```

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+        return ""
+
+    def _pbm_set_config(self, key: str, value: str) -> None:
+        """Runs the charmed-mongodb.pbm config command for the provided key and value."""
+        if self.substrate() == "k8s":
+            config_cmd = [PBM_PATH, "config", "--set", f"{key}={value}"]
+            self.charm.run_pbm_command(config_cmd)
+        else:
+            config_cmd = f'charmed-mongodb.pbm config --set {key}="{value}"'
+            subprocess.check_output(config_cmd, shell=True)
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:398]:
```suggestion
        config_cmd = [PBM_PATH, "config", "--set", f"{key}={value}"]
        self.charm.run_command(config_cmd)
```

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 11]
```diff
@@ -0,0 +1,457 @@
...
+        if not self._check_pbm_installed():
+            return BlockedStatus("pbm not installed.")
+        try:
+            pbm_status = ""
+            if self.substrate() == "k8s":
+                pbm_status = self.charm.run_pbm_command(PBM_STATUS_CMD)
+            else:
+                pbm_status = subprocess.check_output(
+                    "charmed-mongodb.pbm status", shell=True, stderr=subprocess.STDOUT
+                ).decode("utf-8")
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:422]:
```suggestion
            pbm_status = self.charm.run_command(PBM_STATUS_CMD)
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: src/charm.py [hunk 12]
```diff
@@ -291,10 +339,7 @@ def _on_start(self, event) -> None:
     def _relation_changes_handler(self, event) -> None:
         """Handles different relation events and updates MongoDB replica set."""
         self._connect_mongodb_exporter()
-
-        if type(event) is RelationDepartedEvent:
-            if event.departing_unit.name == self.unit.name:
-                self.unit_peer_data.setdefault("unit_departed", "True")
```
Comment (delgod) [inline src/charm.py:297]: hm, will it broke downscaling?

File: src/charm.py [hunk 13]
```diff
@@ -483,11 +565,12 @@ def _init_backup_user(self):
 
         with MongoDBConnection(self.mongodb_config) as mongo:
             # first we must create the necessary roles for the PBM tool
-            logger.debug("creating the backup user roles...")
+            logger.info("creating the backup user roles...")
```
Comment (delgod) [inline src/charm.py:568]:
```suggestion
            logger.info("creating the backup user and roles...")
```

File: src/charm.py [hunk 14]
```diff
@@ -753,15 +837,62 @@ def _connect_mongodb_exporter(self) -> None:
...
+        container.replan()
+
+    def get_backup_service(self) -> ServiceInfo:
+        """Returns the backup service."""
+        container = self.unit.get_container(Config.CONTAINER_NAME)
+        return container.get_service(Config.Backup.SERVICE_NAME)
+
+    def get_container(self) -> Container:
+        """Returns the workload container."""
+        return self.unit.get_container(Config.CONTAINER_NAME)
```
Comment (delgod) [inline src/charm.py:894]: do we really need it public?

**Review by delgod — CHANGES_REQUESTED** [review]

File: src/charm.py (hunk 12 above)
Comment (dmitry-ratushnyy) [reply src/charm.py]: fixed, thank you

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 1 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/mongodb_backups.py]: fixed, thank you

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 2 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/mongodb_backups.py]: Fixed, thank you

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 5 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/mongodb_backups.py]: Fixed, thank you

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 8 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/mongodb_backups.py]: Fixed, thank you

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 9 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/mongodb_backups.py]: Fixed, thank you

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 10 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/mongodb_backups.py]: Fixed, thank you

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 11 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/mongodb_backups.py]: Fixed, thank you

File: src/charm.py (hunk 14 above)
Comment (dmitry-ratushnyy) [reply src/charm.py]: Fixed, thank you

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+                "Sync-ing configurations needs more time, must wait before creating a backup."
+            )
+            return
+        if isinstance(pbm_status, BlockedStatus):
+            event.fail(f"Cannot create backup {pbm_status.message}.")
+            return
+
+        # TODO create backup
+
+    def _on_list_backups_action(self, event) -> None:
```
Comment (dmitry-ratushnyy) [inline lib/charms/mongodb/v0/mongodb_backups.py:147]: This function is copied from VM charm without changes

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+                "Sync-ing configurations needs more time, must wait before listing backups."
+            )
+            return
+        if isinstance(pbm_status, BlockedStatus):
+            event.fail(f"Cannot list backups: {pbm_status.message}.")
+            return
+
+        # TODO list backups
+
+    def _on_restore_action(self, event) -> None:
```
Comment (dmitry-ratushnyy) [inline lib/charms/mongodb/v0/mongodb_backups.py:168]: This function is copied from VM charm without changes

File: lib/charms/data_platform_libs/v0/s3.py
Comment (dmitry-ratushnyy) [inline lib/charms/data_platform_libs/v0/s3.py:1]: This file copied from VM charm without changes

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+                            "waiting to sync s3 configurations."
+                        )
+                        raise ResyncError
+                except ExecError as e:
+                    self.charm.unit.status = BlockedStatus(process_pbm_error(e.stdout))
+
+    def _get_pbm_status(self) -> StatusBase:
+        """Retrieve pbm status."""
+        # TODO VM charm should implement this method
+        self.charm.get_backup_service()
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:338]: I think we can get `ModelError` here. So `try:` block needed

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+        if not self.charm.db_initialised:
+            logger.info("Deferring: set PBM configurations, MongoDB has not yet started.")
+            event.defer()
+            return
+
+        try:
+            # TODO VM charm should implement this  methodx§
+            self.charm.get_backup_service()
+        except ModelError:
+            logger.error("Cannot set PBM configurations, pbm-agent service not found.")
```
Comment (delgod) [inline lib/charms/mongodb/v0/mongodb_backups.py:110]:
```suggestion
            logger.info("Deferring: Cannot set PBM configurations, pbm-agent service not found.")
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 3 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/mongodb_backups.py:211]: Fixed, thank you

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 18 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/mongodb_backups.py:110]: Fixed, thank you

File: lib/charms/operator_libs_linux/v1/snap.py
Comment (dmitry-ratushnyy) [inline lib/charms/operator_libs_linux/v1/snap.py:1]: This file is copied from VM charm as-is

**Review by delgod — APPROVED** [review]

**Review by paulomach — COMMENTED** [review]
could use some more refinements

**Comment by dmitry-ratushnyy** [comment]
> could use some more refinements

for example? 😄

File: lib/charms/mongodb/v0/mongodb_backups.py [hunk 19]
```diff
@@ -0,0 +1,359 @@
...
+        credentials = self.s3_client.get_s3_connection_info()
+        for s3_option, s3_value in credentials.items():
+            if s3_option not in S3_PBM_OPTION_MAP:
+                continue
+
+            pbm_configs[S3_PBM_OPTION_MAP[s3_option]] = s3_value
+        return pbm_configs
+
+    def _resync_config_options(self):
+        """Attempts to sync pbm config options and sets status in case of failure."""
```
Comment (Mehdi-Bendriss) [inline lib/charms/mongodb/v0/mongodb_backups.py:266]: Out of curiosity, what does "syncing pbm config" mean?

File: src/charm.py [hunk 20]
```diff
@@ -281,7 +323,14 @@ def _on_start(self, event) -> None:
...
                 return
 
-        self._connect_mongodb_exporter()
+        try:
+            self._connect_mongodb_exporter()
+        except ChangeError as e:
+            logger.error(
+                "An exception occurred when starting mongodb exporter, error: %s.", str(e)
+            )
+            self.unit.status = BlockedStatus("couldn't start mongodb exporter")
```
Comment (Mehdi-Bendriss) [inline src/charm.py:332]: Should this event be deferred in this case?

File: src/charm.py [hunk 21]
```diff
@@ -753,15 +842,58 @@ def _connect_mongodb_exporter(self) -> None:
...
         # must wait for leader to set URI before connecting
         if not self.get_secret("app", MonitorUser.get_password_key_name()):
             return
+
+        current_service_config = (
+            container.get_plan().to_dict().get("services", {}).get("mongodb_exporter", {})
+        )
+        new_service_config = self._monitor_layer.services.get("mongodb_exporter", {})
+
+        if current_service_config == new_service_config:
```
Comment (Mehdi-Bendriss) [inline src/charm.py:857]: nice

**Review by Mehdi-Bendriss — APPROVED** [review]
Looks good! just minor nits and perhaps some questions

File: src/charm.py (hunk 20 above)
Comment (dmitry-ratushnyy) [reply src/charm.py:332]: I don't think so, this  behaviour is same as in VM charm, but just catches different exception

File: lib/charms/mongodb/v0/mongodb_backups.py (hunk 19 above)
Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v0/mongodb_backups.py:266]: It is  the process of ensuring the PBM configuration is consistent across all nodes in MongoDB cluster.

**Review by delgod — DISMISSED** [review]

**Review by Mehdi-Bendriss — APPROVED** [review]

**Pull request merged**